### DISCUSSION TITLE
Plugin Architecture proposal

### DISCUSSION THREAD

**Discussion by jgbernalp** [perses/perses#2085] [Design Discussions]
# Perses Plugins

## Introduction

Perses plugins aim to extend the functionality of Perses by allowing the creation of new datasources and charts. This modular approach enhances flexibility, enabling users and developers to customize and expand Perses according to their specific needs.

## Types of Plugins

There are two types of plugins in Perses:

1. **[Datasource Plugins](#datasource-plugins)**
2. **[Chart Plugins](#chart-plugins)**

### Datasource Plugins

Datasource plugins fetch data from various sources. Each datasource plugin consists of:

- **Backend**

  - `model.cue` file: Defines the datasource model.
  - `migrate.cue` file: Specifies migration procedures from other formats to the datasource model.

- **Frontend**
  - `fetch.ts` file: Utilizes the Perses SDK and plugin configuration to fetch data and parse it into a DataFrame.
  - `DatasourceEditor.ts` file: Contains a React component for editing the datasource model fields.
  - `QueryEditor.ts` file: Contains a React component for editing the query used to fetch data.

### Chart Plugins

Chart plugins render data in various formats. Each chart plugin consists of:

- **Frontend**
  - `Chart.tsx` file: Contains a React component that renders the chart. It receives data as props in the form of a DataFrame or DataFrame Array. While any charting library can be used, ECharts is recommended as it is included with the Plugin runtime.

## Perses SDK

The Perses SDK provides utilities to facilitate the creation of plugins and the consumption of the Perses API. It allows developers to:

- Fetch data from datasources, managing caching and refetching.
- Interact with the Perses API to create, update, and delete resources such as datasources, dashboards, variables, and projects.
- Define core models used when implementing a plugin.
- Get variable values, such as time range or custom variables.

## Plugin Runtime

The plugin runtime includes utilities that enable the Perses UI or other frontends to load plugins at runtime. It uses Module Federation to load plugins and expose them to consumers.

## Diagram

<img width="1217" alt="Screenshot 2024-06-19 at 17 56 02" src="https://github.com/perses/perses/assets/5461414/2889faf1-fa7a-4bb2-91db-e79ab8ec25a6">

## Plugin Development

Developers can create plugins by following the steps below:

1. Create a plugin using `npx create perses-plugin`. Use the options provided to select the type of plugin.
2. Run Perses locally in `plugin-dev` mode to implement, debug and test the plugin.
3. Package the plugin for installation using `npm run package:plugin`.
4. Run Perses locally and install the plugin.
5. Test the plugin in Perses.
6. Submit a pull request to the Perses repository to add the plugin to the catalog.

## Migration Plan

The migration plan involves transitioning current datasources and charts to plugins. This will allow users to define their own datasources and charts and connect them using a common DataFrame format. Existing datasources and charts will be migrated and available as default plugins.

### Migration Steps

- [x] Use rspack to build ui/app #2002
- [x] Create plugin bootstrap tool https://github.com/perses/poc-plugin-v2/pull/3
- [x] Complete Proof of Concept (PoC) for loading Chart Plugins using Module Federation in the Perses UI, initially targeting TimeSeries. https://github.com/perses/perses/pull/2091
- [ ] Decouple data fetching, chart theming, and core models into the new Perses SDK. This will enable plugins to operate independently of the main Perses codebase.
- [ ] Migrate the current frontend for datasources to plugins.https://github.com/perses/plugins/pull/16
- [ ] Migrate the current frontend for charts to plugins. https://github.com/perses/plugins/pull/16
- [ ] Migrate the current backend for datasources to plugins.
- [ ] Load plugins dynamically on Perses https://github.com/perses/perses/pull/2514
- [ ] Merge plugin-runtime and plugin-system packages

**Comment by AntoineThebaud** [comment]
Thanks for the well-wrapped proposal 🙏 

About data frames, if I'm not mistaken you are going with the 2nd design described here right https://github.com/perses/perses/discussions/1922? WDYT about sticking instead to core-defined formats? (https://github.com/perses/perses/discussions/1922#discussioncomment-9828738)

Also, what about Variable & Query plugins? You didn't mention them - or do you consider them as parts of the Datasource plugins?

**Reply by andreasgerstmayr** [reply]
Signals like traces have a [rich data format](https://github.com/open-telemetry/opentelemetry-proto/blob/main/examples/trace.json). Some data sources (Tempo) may provide data in multiple formats based on the query: either a SearchResult (for TraceQL queries) or an entire Trace (if the query is a valid trace ID).

I'm +1 on the `DataFrame` approach, i.e. panels should not be coupled with the data source, but instead getting the data passed as props. For the data format though, I'd stick to our core-defined formats ([time series](https://github.com/perses/perses/blob/main/ui/core/src/model/time-series-data.ts), [traces](https://github.com/perses/perses/blob/main/ui/core/src/model/trace-data.ts)).

**Reply by jgbernalp** [reply]
@AntoineThebaud Query plugins were not necessary in the only-DataFrame approach as the initial thought was that all can be converted into it 😓 . But I see now the complexities with other formats, and the errors that might arise while mapping. Leaning to that. I'll give another pass to this, including the QueryPlugins which at the end are in charge of mapping the data to a core-defined format. I'll include Variables as well.

**Comment by Nexucis** [comment]
### BACKEND API

At some points the frontend will need to get the list of the available plugins. The backend can provide the list through an HTTP endpoint. Probably `/api/v1/plugins`. 

This endpoint will return a the list of the available plugin in a JSON format. It will include:
- the name
- the version, 
- the type (chart / datasources / query / variable)
- other things ... ?

Some questions from my side about what the backend should do: 

- Should the backend be able to download the archive ?
- Should the backend be able to unzip the archive and put everything in a defined path (from config)
- How can we define the type of a plugin (is it a chart, is it a query) ..etc. ?
- If there is already a list of plugin already installed, and a list of archive, should the backend erase what has been already installed in favor of the archive to be unzip ?

### Cuelang files

We will have to move the cuelang files also, these files will have to be part of the plugin archive. And as we are using them for the SDK, that means we will have to package them to be able to use them.

Currently the cuelang files depend on common libraries. I guess we will have to move them as well to the plugin repository. Just to mention that will need somehow to install these common libs in the Perses backend because otherwise we won't be able to use the cue files. Maybe I am wrong, but I think this part will be tricky.

As a side note, I believe the golang files associated to the plugin will have to move as well. Which means each plugin will have a golang mod files.

**Reply by AntoineThebaud** [reply]
I was thinking again about this backend part of the plugins.
- So yes first for sure, each plugin in perses/perses-plugins should define its own `cue.mod/module.cue`.
- To me the `common` cue package could either remain in perses/perses or move to perses/perses-plugins (where it would have its own `cue.mod` like the plugins). Both make sense somehow.
- Then regarding how each plugin would inherite/retrieve the common package: when we discussed last time we were talking about leveraging on the new native dependency management of CUE, however this seems tricky to do in the current case (I can elaborate on this part if you're interested..).
Since on the other hand the frontend side is progressing well, we should consider implementing this without this native dep mgt on the backend side for now, in order to not block the move to the new plugin architecture. 
- We already discussed about adding a new command to `percli` to bootstrap a new plugin project. @jgbernalp already did an npm utility to generate the skeleton of the frontend part only, so I think we could have a new `percli` command that would:
  - wrap the call to this `npm create perses-plugin`,
  - run `cue mod init` to generate the "structure" of the backend part,
  - "vendor" the `common` cue package dependency: retrieve (download) the `common` cue package and store it under `cue.mod/pkg` 
-> very similar to what we do already with the `percli dac setup` command for DaC. This is the part we would replace later on by proper dependency management.
  - (optional) based on a flag, generate the skeleton of a Go project if the plugin developer wish to integrate with the Go SDK.

NB: to be able to only update the common package dependency in case a new version is available, maybe there should be a dedicated command for that? to not run the npm part every time.

**Comment by Nexucis** [comment]
This discussion can be closed as the new plugin system has been implemented and we are going to release it.

Documentation and blog post will give more details.
